File: tools/split_genres.py
# This script splits genres that contain ", " into multiple genres so you can more easily search for a single tag.
import base64

import requests

# Configuration constants

# PLEASE MAKE A BACKUP BEFORE RUNNING THIS SCRIPT!

ABS_HOST = ""  # AudiobookShelf Host URL
API_KEY = ""
LIBRARY_IDS = []  # Leave empty to process all libraries
SKIP_GENRES = ['Mystery, Thriller & Suspense']  # Genres to skip


###### Code ######

def get_all_genres():
    response = requests.get(
        f"{ABS_HOST}/api/genres",
        params={"token": API_KEY}
    )
    response.raise_for_status()
    return response.json()['genres']


# Returns all genres that include ", "
def get_multi_genres(genres):
    multi_genres = []
    for genre in genres:
        if ", " in genre:
            if genre not in SKIP_GENRES:
                multi_genres.append(genre)
    return multi_genres


def get_all_books_for_genre(genre: str, id):
    base64_genre = genre.encode('utf-8')
    url_encoded_genre = base64.urlsafe_b64encode(base64_genre).decode('utf-8')

    response = requests.get(
        f"{ABS_HOST}/api/libraries/{id}/items",
        params={
            "limit": 99999,
            "filter": f"genres.{url_encoded_genre}",
            "token": API_KEY
        }
    )
    response.raise_for_status()
    return response.json()['results']


def append_genre_to_book(book, genre):
    genres = genre.split(", ")
    book_genres = book['media']['metadata']['genres']

    # Join the two lists and remove duplicates
    book_genres = list(set(book_genres + genres))
    # Remove the genre from the book
    book_genres.remove(genre)

    book_id = book['id']

    patch_body = {"metadata": {"genres": book_genres}}

    response = requests.patch(
        f"{ABS_HOST}/api/items/{book_id}/media",
        params={"token": API_KEY},
        json=patch_body
    )
    response.raise_for_status()


def get_all_libraries():
    response = requests.get(
        f"{ABS_HOST}/api/libraries",
        params={"token": API_KEY}
    )
    response.raise_for_status()
    return response.json()['libraries']


if __name__ == "__main__":

    if len(LIBRARY_IDS) == 0:
        libraries = get_all_libraries()
        for library in libraries:
            LIBRARY_IDS.append(library['id'])

    print(f"Processing {len(LIBRARY_IDS)} libraries")
    print(LIBRARY_IDS)
    print("----------------------")
    multi_genres = get_multi_genres(get_all_genres())

    for genre in multi_genres:
        print(f"Processing genre {genre}", end="\n\n")
        book_titles_overall = []
        for library_id in LIBRARY_IDS:
            books = get_all_books_for_genre(genre, library_id)
            book_titles = [book['media']['metadata']['title'] for book in books]
            book_titles_overall += book_titles
            for book in books:
                try:
                    append_genre_to_book(book, genre)
                except Exception as e:
                    print(f"Error processing book {book['id']}: {e}")
        print(f"Processed {len(book_titles_overall)} books for genre {genre}: {book_titles_overall}", end="\n\n")
        print("----------------------", end="\n")

File: web/src/tools/splitGenres.ts
import { useApi } from '@/composables/useApi'
import type { ToolResult } from '@/types/tool'

async function getAllBooksForGenre(genre: string, libraryId: string) {
  const { get, addLog } = useApi()
  try {
    const base64Genre = btoa(unescape(encodeURIComponent(genre))).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');

    const response = await get(`/api/libraries/${libraryId}/items?filter=genres.${base64Genre}`)
    const bookCount = response.data.results.length
    addLog(`Found ${bookCount} books for genre "${genre}" in library ${libraryId}`)
    return response.data.results || []
  } catch (error) {
    addLog(`Error fetching books for genre "${genre}" in library ${libraryId}`)
    console.error(`Error fetching books for genre ${genre} in library ${libraryId}:`, error)
    return []
  }
}

async function appendGenreToBook(book: any, genre: string, delimiter: string, type: 'genres' | 'tags') {
  const { patch, addLog } = useApi()
  const bookId = book.id
  const genres = genre.split(delimiter).map((g: string) => g.trim())

  try {
    await patch(`/api/items/${bookId}/media`, {
      metadata: {
        [type]: [...new Set([...book.media.metadata[type], ...genres].filter(g => g !== genre))]
      }
    })
    addLog(`Updated book: ${book.media.metadata.title}`)
  } catch (error) {
    addLog(`Error updating book: ${book.media.metadata.title}`)
    console.error(`Error appending genre ${genre} to book ${bookId}:`, error)
  }
}

export async function executeSplitGenres(formData: Record<string, any>): Promise<ToolResult> {
  const { get, post, addLog } = useApi()
  
  try {
    const { type, libraryIds, delimiter } = formData

    addLog('Starting split genres operation...')
    console.log('Executing split genres with formData:', formData)

    const libraryResponse = await get('/api/libraries')

    const processableLibraries = [];

    for (const library of libraryResponse.data.libraries) {
      if (libraryIds.length === 0 || libraryIds.includes(library.id)) {
        processableLibraries.push(library.id)
      }
    }

    const libraryMessage = `Processing ${processableLibraries.length} libraries`
    addLog(libraryMessage)

    const genres = (await get(`/api/${type}`)).data.genres || [];

    const multiGenres = [];

    for (const genre of genres) {
      if (genre.split(delimiter).length > 1 && !formData.skip.includes(genre)) {
        multiGenres.push(genre);
      }
    }

    addLog(`Found ${multiGenres.length} genres to split`)

     for (const genre of multiGenres) {
       const genreMessage = `Processing genre: ${genre}`
       addLog(genreMessage)
       
       const bookTitlesOverall: string[] = [];
       for (const libraryId of processableLibraries) {
         const books = await getAllBooksForGenre(genre, libraryId);
         const bookTitles = books.map((book: any) => book.media.metadata.title);
         bookTitlesOverall.push(...bookTitles);
         for (const book of books) {
           try {
             await appendGenreToBook(book, genre, delimiter, type);
           } catch (e) {
             console.log(`Error processing book ${book.id}: ${e}`);
           }
         }
       }
       addLog(`Processed ${bookTitlesOverall.length} books for genre ${genre}: ${bookTitlesOverall.join(', ')}`);
     }

    const successMessage = `${type} split operation completed successfully`
    addLog(successMessage)

    return {
      success: true,
      message: successMessage,
      timestamp: new Date().toISOString()
    }

  } catch (error: any) {
    const errorMessage = 'Failed to split genres'
    addLog(errorMessage)
    return {
      success: false,
      message: errorMessage,
      error: error.message || 'Unknown error',
      timestamp: new Date().toISOString()
    }
  }
}